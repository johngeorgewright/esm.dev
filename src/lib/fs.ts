export async function readJSONFile(filename: string) {
  return JSON.parse(await Deno.readTextFile(filename))
}

export async function writeJSONFile(filename: string, data: unknown) {
  await Deno.writeTextFile(filename, JSON.stringify(data, null, 2))
}
