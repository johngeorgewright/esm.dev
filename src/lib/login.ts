import { spawn } from 'node:child_process'

export function login(registry: string): Promise<number> {
  return new Promise((resolve, reject) => {
    const child = spawn('npm', ['login', '--registry', registry, '--quiet'])

    child.stderr.on('data', (d) => console.error(d.toString()))

    child.stdout.on('data', (d) => {
      const data = d.toString()
      switch (true) {
        case /username/i.test(data):
          child.stdin.write('esm.dev\n')
          break
        case /password/i.test(data):
          child.stdin.write('esm.dev\n')
          break
        case /email/i.test(data):
          child.stdin.write('esm.dev@esm.dev.com')
          break
        case /logged in as/i.test(data):
          child.stdin.end()
          break
      }
    })

    child.on('exit', (code) => code && reject(code) || resolve(code ?? 0))
  })
}
