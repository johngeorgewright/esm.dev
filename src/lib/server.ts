import { queue } from './queue.ts'

export async function serve(
  port: number,
  esmOrigin: string,
): Promise<() => Promise<void>> {
  const { promise, resolve } = Promise.withResolvers<void>()

  const server = Deno.serve({
    port,
    onListen(netAddr) {
      console.info('ESM proxy server listining on', netAddr)
      resolve()
    },
  }, (request) =>
    queue(() => {
      const { pathname, search } = new URL(request.url)
      const url = new URL('.' + pathname, esmOrigin)
      url.search = search

      const headers = new Headers(request.headers)
      headers.set('Host', url.hostname)

      return fetch(url, {
        method: request.method,
        headers,
        body: request.body,
        redirect: 'manual',
      })
    }).catch((error) =>
      new Response(JSON.stringify(error), {
        headers: {
          'Content-Type': 'application/json; charset=utf-8',
        },
        status: 500,
      })
    ))

  await promise

  return async () => {
    console.info('closing the server')
    await server.shutdown()
  }
}
