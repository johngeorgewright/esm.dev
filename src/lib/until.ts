import { setTimeout } from 'node:timers/promises'

export async function until({
  interval,
  timeout,
  try: Try,
}: {
  interval: number
  timeout: number
  try(signal: AbortSignal): Promise<boolean>
}): Promise<boolean> {
  const signal = AbortSignal.timeout(timeout)
  while (!signal.aborted) {
    try {
      if (await Try(signal)) return true
    } catch (error) {}
    try {
      await setTimeout(interval, null, { signal })
    } catch (error) {}
  }
  return false
}

export async function waitForEndpoint({
  interval = 300,
  timeout = 10_000,
  endpoint,
}: {
  interval?: number
  timeout?: number
  endpoint: string
}): Promise<void> {
  if (
    !(await until({
      interval,
      timeout,
      async try(signal) {
        const response = await fetch(endpoint, { signal })
        return response.ok
      },
    }))
  )
    throw new EndpointUnavailableError()
}

export class EndpointUnavailableError extends Error {
  constructor() {
    super('Endpoint not available')
  }
}
