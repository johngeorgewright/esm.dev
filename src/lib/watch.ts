import { watch as fsWatch, watchFile } from 'node:fs'
import { republish } from './republish.js'
import { getPackageMeta } from './getPackageMeta.js'
import { readFile, writeFile } from 'node:fs/promises'
import { createHash } from 'node:crypto'
import { glob } from 'glob'
import { queue, queuedDebounce } from './queue.js'
import { getWatchIgnorer, type Ignorer } from './watchIgnoreList.js'

export async function watch(
  packagePath: string,
  opts: { registry: string; esmStoragePath: string; legacyMethod?: boolean },
): Promise<() => void> {
  const {
    name,
    packageRoot,
    private: prvte,
  } = await getPackageMeta(packagePath)

  if (prvte) {
    console.info(`${name} is a private package... ignoring`)
    return () => {}
  }

  console.info('Watching', packageRoot)

  const abortController = new AbortController()
  const signal = abortController.signal

  const republishPackage = () =>
    republish(packageRoot, opts).catch(console.error)

  const debounceRepublish = queuedDebounce(
    async (filename: string = '') => {
      console.info(`Change detected at ${packageRoot}/${filename}`)
      await republishPackage()
    },
    1_000,
    signal,
  )

  await queue(signal, republishPackage)

  return opts.legacyMethod
    ? legacyWatch(abortController, packageRoot, debounceRepublish)
    : modernWatch(abortController, packageRoot, debounceRepublish)
}

async function modernWatch(
  abortController: AbortController,
  dirname: string,
  debounceRepublish: (filename?: string | undefined) => Promise<void>,
) {
  const ignorer = await getWatchIgnorer(dirname)
  const watcher = fsWatch(
    dirname,
    { recursive: true },
    (_, filename) =>
      (filename && ignorer.ignores(filename)) ||
      debounceRepublish(filename ?? ''),
  )
  return () => {
    console.info('Stop watching')
    watcher.close()
    abortController.abort()
  }
}

async function legacyWatch(
  abortController: AbortController,
  dirname: string,
  debounceRepublish: (filename?: string | undefined) => Promise<void>,
) {
  const ignorer = await getWatchIgnorer(dirname)
  const filename = await hashDirectory(dirname, ignorer)
  const watcher = watchFile(filename, () => debounceRepublish())
  let timer: NodeJS.Timeout | undefined
  beginTimer()
  return () => {
    console.info('Stop watching')
    watcher.unref()
    clearTimeout(timer)
    abortController.abort()
  }
  function beginTimer() {
    timer = setTimeout(
      () =>
        hashDirectory(dirname, ignorer)
          .catch(console.error)
          .finally(beginTimer),
      1_000,
    )
  }
}

async function hashDirectory(dirname: string, ignorer: Ignorer) {
  const filename = `/tmp/${dirname.replace(/\//g, '-')}.hash.txt`
  const files = ignorer.filter(await glob(`${dirname}/**/*`, { nodir: true }))
  const hashes = await Promise.all(
    files.map(async (file) => {
      const contents = await readFile(file, 'utf-8')
      return createHash('sha256').update(contents).digest('hex')
    }),
  )
  const newHash = hashes.join('')

  let previousHash = ''

  try {
    previousHash = await readFile(filename, 'utf-8')
  } catch (error: any) {
    if (error.code !== 'ENOENT') throw error
  }

  if (newHash !== previousHash) await writeFile(filename, newHash)
  return filename
}
