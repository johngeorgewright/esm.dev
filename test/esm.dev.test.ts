import { assertSnapshot } from 'jsr:@std/testing/snapshot'
import { readFile, writeFile } from 'node:fs/promises'
import { setTimeout } from 'node:timers/promises'
import { serve } from '../src/lib/server.ts'
import { watch } from '../src/lib/watch.ts'
import { login } from '../src/lib/login.ts'
import { waitForEndpoint } from '../src/lib/until.ts'
import { esmOrigin, esmStoragePath, port, registry } from './constants.ts'

Deno.test('dev environment', async (t) => {
  await t.step('setup', async () => {
    await waitForEndpoint({ endpoint: registry })
    await login(registry)
  })

  await t.step('access', async (t) => {
    await using _disposable = await start()

    for (
      const { endpoint, pckg } of [
        {
          endpoint: '/@esm.dev/package-1@0.0.1/es2022/package-1.mjs',
          pckg: 'pacakge-1',
        },
        { endpoint: '/@esm.dev/package-2', pckg: 'package-2' },
      ]
    ) {
      await t.step(`ESM server retreives watched ${pckg}`, async (t) => {
        const response = await fetch(`http://localhost:${port}${endpoint}`)
        await assertSnapshot(t, await response.text())
      })
    }
  })

  for (
    const { legacyMethod, interval, name } of [
      { legacyMethod: false, name: 'modern', interval: 100 },
      { legacyMethod: true, name: 'legacy', interval: 3_000 },
    ]
  ) {
    await t.step(`watching with ${name} method`, async (t) => {
      await using _disposable = await start(legacyMethod)
      await using _mainExport = await changeMainExportDisposable(interval)
      const response = await fetch(
        `http://localhost:${port}/@esm.dev/package-1@0.0.1/es2022/package-1.mjs`,
      )
      await assertSnapshot(t, await response.text())
    })
  }
})

async function start(legacyMethod?: boolean): Promise<AsyncDisposable> {
  const disposableStack = new AsyncDisposableStack()
  disposableStack.adopt(await serve(port, esmOrigin), (stop) => stop())
  disposableStack.adopt(
    await watch('test/packages/package-1', {
      registry,
      esmStoragePath,
      legacyMethod,
    }),
    (stop) => stop(),
  )
  disposableStack.adopt(
    await watch('test/packages/package-2', {
      registry,
      esmStoragePath,
      legacyMethod,
    }),
    (stop) => stop(),
  )
  return disposableStack
}

async function changeMainExportDisposable(
  interval: number,
): Promise<AsyncDisposable> {
  await changeMainExport('./src/foos.ts', interval)
  return {
    [Symbol.asyncDispose]: () => changeMainExport('./src/foo.ts', interval),
  }
}

async function changeMainExport(to: string, interval: number) {
  const filename = 'test/packages/package-1/package.json'
  const json = JSON.parse(await readFile(filename, 'utf-8'))
  json.exports['.'] = to
  await writeFile(filename, JSON.stringify(json, null, 2))
  await setTimeout(interval)
}
