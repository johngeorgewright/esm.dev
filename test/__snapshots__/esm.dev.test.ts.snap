export const snapshot = {};

snapshot[`dev environment > access > ESM server retreives watched pacakge-1 1`] = `
'/* esm.sh - @esm.dev/package-1@0.0.1 */
function o(){return"foo"}export{o as foo};
//# sourceMappingURL=package-1.mjs.map'
`;

snapshot[`dev environment > access > ESM server retreives watched package-2 1`] = `
'/* esm.sh - @esm.dev/package-2@0.0.1 */
export * from "/@esm.dev/package-2@0.0.1/denonext/package-2.mjs";
'
`;

snapshot[`dev environment > watching with modern method 1`] = `
'/* esm.sh - @esm.dev/package-1@0.0.1 */
function o(){return"foo"}export{o as foo};
//# sourceMappingURL=package-1.mjs.map'
`;

snapshot[`dev environment > watching with legacy method 1`] = `
'/* esm.sh - @esm.dev/package-1@0.0.1 */
function o(){return"foo"}export{o as foo};
//# sourceMappingURL=package-1.mjs.map'
`;
